src/backend/distributed/executor/multi_client_executor.c

```diff
@@ -315,6 +315,53 @@ MultiClientConnectionUp(int32 connectionId)
 }
 
 
+/* MultiClientExecute synchronously executes a query over the given connection. */
+bool
+MultiClientExecute(int32 connectionId, const char *query, void **queryResult,
+				   int *rowCount, int *columnCount)
+{
+	bool querySent = false;
+	bool queryReady = false;
+	bool queryOK = false;
+	WaitInfo *waitInfo = NULL;
+
+	querySent = MultiClientSendQuery(connectionId, query);
+	if (!querySent)
+	{
+		return false;
+	}
+
+	waitInfo = MultiClientCreateWaitInfo(1);
+
+	while (!queryReady)
+	{
+		ResultStatus resultStatus = MultiClientResultStatus(connectionId);
+		if (resultStatus == CLIENT_RESULT_READY)
+		{
+			queryReady = true;
+		}
+		else if (resultStatus == CLIENT_RESULT_BUSY)
+		{
+			/* wait for results, errors, or interrupts */
+			MultiClientResetWaitInfo(waitInfo);
+			MultiClientRegisterWait(waitInfo, TASK_STATUS_SOCKET_READ, connectionId);
+			MultiClientWait(waitInfo);
```

Comment:
I was looking for a way to handle cancellation in this function, though it's not  relevant for master_append_table_to_shard. I could remove the while loop and let it block in MultiClientQueryResult, though I feel like that warrants a /* TODO: handle cancellation */. 

I suppose that if we fix #671 by removing the CHECK_FOR_INTERRUPTS then we would have to check for cancellation explicitly here if we want to handle it. That's something I didn't consider while writing this.

I'd be ok ditching the loop for now, what do you think?